notes/stv/20191209-openstack-deployment.txt

```diff
@@ -0,0 +1,919 @@
+#   <meta:licence>
+#     Copyright (c) 2019, ROE (http://www.roe.ac.uk/)
+#
+#     This information is free software: you can redistribute it and/or modify
+#     it under the terms of the GNU General Public License as published by
+#     the Free Software Foundation, either version 3 of the License, or
+#     (at your option) any later version.
+#
+#     This information is distributed in the hope that it will be useful,
+#     but WITHOUT ANY WARRANTY; without even the implied warranty of
+#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#     GNU General Public License for more details.
+#  
+#     You should have received a copy of the GNU General Public License
+#     along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#   </meta:licence>
+# </meta:header>
+#
+#
+
+
+
+################################################################
+##  Setup Zepellin/Yarn/Spark/Hadoop Cluster                  ##
+################################################################
+
+
+# Read up and try to follow
+# https://docs.hpc.cam.ac.uk/cloud-training/#16
+
+
+#-------------------------------------------------------------------------
+# GUI Steps to Setup Networks, Security Groups & Instances
+#-------------------------------------------------------------------------
+
+
+## Step 1:
+
+Create Private Network 
+
+   Name: stv-dev-network
+
+   Create Subnet
+      Name - stv-dev-subnet 
+      CIDR - 10.0.0.0/24
+      DNS Name Servers - 131.111.8.42 131.111.12.20      
+
+
+
+## Step 2:
+
+Create Router, connected to Internet
+ 
+   Name: stv-dev-router
+   External Network: Internet
+
+   Create Interface
+      Subnet - stv-dev-network
+
+
+## Step 3
+
+Create Security Groups (based on training presentation)
+
+external-bastion: 22 / ICMP
+internal-bastion: default
+internal-webserver: 22 - internal-bastion
+external-webserver: 80
+stv-dev-zeppelin : 8080
+stv-dev-master: *
+stv-dev-worker: * - stv-dev-master
+
+
+
+## Step 4:
+
+Create Bastion Instance (Gateway)
+    Fedora-30-1.2
+    general.v1.tiny	
+    external-bastion + internal-bastion
+
+
+## Step 5:
+
+Create Master node
+    Fedora-30-1.2
+    general.v1.medium	
+    internal-bastion + internal-webserver + stv-dev-master
+
+
+
+## Step 6:
+
+Create 6 Worker Nodes with 500 Gb volumes
+    Fedora-30-1.2
+    general.v1.large	
+    (external-bastion + internal-webserver + stv-dev-worker)
+
+
+
+## Step 7:
+
+Create Zeppelin Node 
+(external-bastion + internal-webserver + stv-dev-zeppelin)
+
+
+
+
+
+stv-dev-gateway-ip=
+stv-dev-worker-1-ip=
+stv-dev-worker-2-ip=
+stv-dev-worker-3-ip=
+stv-dev-master-ip=
+stv-dev-zeppelin-ip=
+
+
+
+
+#-------------------------------------------------------------------------
+# Setup .ssh config locally
+#-------------------------------------------------------------------------
+
+~/.ssh/config
+..
+
+
+Host stv-dev-gateway
+    User centos
+    IdentityFile ~/.ssh/stv-master
+    Hostname ${stv-dev-gateway-ip:?}
```

Comment:
This bash variable substitution wouldn't work in ssh so how was it resolved ?

Reply:
The placeholders were added before committing the notes, mainly to hide the public IPs in github. I've updated the notes to change dashes to underscores.

Reply:
The ${something:?} syntax still looks like a bash variable substitution.
If you want to just show a manual substitution, just {something} without the $ and :? might be clearer.
I wouldn't worry too much about revealing public IP addresses, the bad guys already know them.

Reply:
Ah - ok, my bad. I was looking at the four line fragment above, and didn't realise what the 'Outdated' label meant. It meant the fragment was out of date and I needed to go look at the latest version of the file.

Reply:
Adding the `cat << EOF >> ~/.ssh/config` makes it clear how the substitution happened.